src/Server/Query.elm

```diff
@@ -29,28 +29,62 @@ type alias Errors =
     Dict FieldName ErrorMessage
 
 
-parse : Db -> Query.Parser (Result Errors Inputs.Query)
-parse db =
-    Query.map Inputs.Query
-        (massParser "mass")
-        |> apply (materialParser "material")
-        |> apply (productParser "product")
-        |> apply (countriesParser "countries")
-        |> apply (ratioParser "dyeingWeighting")
-        |> apply (ratioParser "airTransportRatio")
-        |> apply (ratioParser "recycledRatio")
-        |> apply customCountryMixesParser
-        |> Query.map (validateQuery db)
-
-
-
--- Parsers
+succeed : a -> Query.Parser a
+succeed =
+    -- Kind of a hack: we don't have access to the Query.Parser constructor, so
+    -- we use Query.custom to wrap our thing in a Query.Parser
+    always >> Query.custom ""
 
 
-apply : Query.Parser a -> Query.Parser (a -> b) -> Query.Parser b
+parse : Db -> Query.Parser (Result Errors Inputs.Query)
+parse db =
+    succeed (Ok Inputs.Query)
+        |> apply (massParser "mass")
+        |> apply (materialParser "material" db.materials)
+        |> apply (productParser "product" db.products)
+        |> apply (countryParser "countryFabric" db.countries)
+        |> apply (countryParser "countryDyeing" db.countries)
+        |> apply (countryParser "countryMaking" db.countries)
+        |> apply (maybeRatioParser "dyeingWeighting")
+        |> apply (maybeRatioParser "airTransportRatio")
+        |> apply (maybeRatioParser "recycledRatio")
+        |> apply (Query.map Ok customCountryMixesParser)
+
+
+toErrors : Result ( FieldName, String ) a -> Result Errors a
+toErrors result =
+    Result.mapError (\( key, errorMessage ) -> Dict.singleton key errorMessage) result
```

Comment:
I find this form less clear because it doesn't have the argument next to `toErrors` (so you have to mentally parse the signature), but I remember that this is the style used throughout the codebase, so i'll conform to it :+1: